src/rust/rust-proto-new/src/protocol/status.rs

```diff
@@ -0,0 +1,363 @@
+            Code::OutOfRange => "Operation was attempted past the valid range",
+            Code::Unimplemented => "Operation is not implemented or not supported",
+            Code::Internal => "Internal error",
+            Code::Unavailable => "The service is currently unavailable",
+            Code::DataLoss => "Unrecoverable data loss or corruption",
+            Code::Unauthenticated => "The request does not have valid authentication credentials",
+        }
+    }
+}
+
+impl std::fmt::Display for Code {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(self.description(), f)
+    }
+}
+
+impl From<Code> for GrpcCode {
+    fn from(code: Code) -> Self {
+        match code {
+            Code::Ok => GrpcCode::Ok,
+            Code::Cancelled => GrpcCode::Cancelled,
+            Code::Unknown => GrpcCode::Unknown,
+            Code::InvalidArgument => GrpcCode::InvalidArgument,
+            Code::DeadlineExceeded => GrpcCode::DeadlineExceeded,
+            Code::NotFound => GrpcCode::NotFound,
+            Code::AlreadyExists => GrpcCode::AlreadyExists,
+            Code::PermissionDenied => GrpcCode::PermissionDenied,
+            Code::ResourceExhausted => GrpcCode::ResourceExhausted,
+            Code::FailedPrecondition => GrpcCode::FailedPrecondition,
+            Code::Aborted => GrpcCode::Aborted,
+            Code::OutOfRange => GrpcCode::OutOfRange,
+            Code::Unimplemented => GrpcCode::Unimplemented,
+            Code::Internal => GrpcCode::Internal,
+            Code::Unavailable => GrpcCode::Unavailable,
+            Code::DataLoss => GrpcCode::DataLoss,
+            Code::Unauthenticated => GrpcCode::Unauthenticated,
+        }
+    }
+}
+
+impl From<GrpcCode> for Code {
+    fn from(code: GrpcCode) -> Self {
+        match code {
+            GrpcCode::Ok => Code::Ok,
+            GrpcCode::Cancelled => Code::Cancelled,
+            GrpcCode::Unknown => Code::Unknown,
+            GrpcCode::InvalidArgument => Code::InvalidArgument,
+            GrpcCode::DeadlineExceeded => Code::DeadlineExceeded,
+            GrpcCode::NotFound => Code::NotFound,
+            GrpcCode::AlreadyExists => Code::AlreadyExists,
+            GrpcCode::PermissionDenied => Code::PermissionDenied,
+            GrpcCode::ResourceExhausted => Code::ResourceExhausted,
+            GrpcCode::FailedPrecondition => Code::FailedPrecondition,
+            GrpcCode::Aborted => Code::Aborted,
+            GrpcCode::OutOfRange => Code::OutOfRange,
+            GrpcCode::Unimplemented => Code::Unimplemented,
+            GrpcCode::Internal => Code::Internal,
+            GrpcCode::Unavailable => Code::Unavailable,
+            GrpcCode::DataLoss => Code::DataLoss,
+            GrpcCode::Unauthenticated => Code::Unauthenticated,
+        }
+    }
+}
+
+pub struct Status {
+    /// The gRPC status code, found in the `grpc-status` header.
+    code: Code,
```

Comment:
Why is a `Status` composed of a `code` and a `message`? Does this give us some advantage over an enumeration? It feels to me like an `enum Status` would be more ergonomic to callers.

Reply:
Structs are much easier to extend. If I want to add a field to an enum I have to add that field to every single variant. A struct wrapping an enum is a fairly common pattern for this reason.

Reply:
Do we expect we'll need to add more fields?

Reply:
Possibly. The tonic::Status includes the source rust error, which lets it impl Error, as well as as an http metadata map.